src/ValueObject/DriverOptions.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace MongoDB\ValueObject;
+
+use MongoDB\Builder\BuilderEncoder;
+use MongoDB\Client;
+use MongoDB\Codec\Encoder;
+use MongoDB\Exception\InvalidArgumentException;
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+
+use function array_diff_key;
+use function array_filter;
+use function is_array;
+use function is_string;
+use function sprintf;
+use function trim;
+
+/** @internal */
+final class DriverOptions
+{
+    private const KEY_TYPE_MAP = 'typeMap';
+    private const KEY_BUILDER_ENCODER = 'builderEncoder';
+    private const KEY_AUTO_ENCRYPTION = 'autoEncryption';
+    private const KEY_DRIVER = 'driver';
+    private const DEFAULT_TYPE_MAP = [
+        'array' => BSONArray::class,
+        'document' => BSONDocument::class,
+        'root' => BSONDocument::class,
+    ];
+
+    private const HANDSHAKE_SEPARATOR = '/';
+
+    private function __construct(
+        public readonly array $typeMap,
+        public readonly Encoder $builderEncoder,
+        public readonly array $autoEncryption,
+        public readonly array $driver,
+        private readonly array $miscOptions,
+    ) {
+    }
+
+    public function toArray(): array
+    {
+        return array_filter([
+            'typeMap' => $this->typeMap,
+            'builderEncoder' => $this->builderEncoder,
+            'autoEncryption' => $this->autoEncryption,
+            'driver' => $this->driver,
+        ] + $this->miscOptions);
+    }
+
+    public static function fromArray(array $options): self
+    {
+        $options += ['typeMap' => self::DEFAULT_TYPE_MAP];
+
+        self::ensureValidArrayOptions($options);
+
+        $autoEncryption = match (isset($options[self::KEY_AUTO_ENCRYPTION])) {
+            true => AutoEncryptionOptions::fromArray($options[self::KEY_AUTO_ENCRYPTION])->toArray(),
+            false => []
+        };
```

Comment:
I find the usage of `match` abusive here.
```suggestion
        $autoEncryption = null;
        if (isset($options[self::KEY_AUTO_ENCRYPTION])) {
            $autoEncryption = AutoEncryptionOptions::fromArray($options[self::KEY_AUTO_ENCRYPTION])->toArray();
        }
```